### PR TITLE
feat(restore): snapshot bootstrap + genesis-baked validators + consensus shard enforcement

### PR DESCRIPTION
**Pull request by tcsenpai**
## Summary

Three coordinated epics that make a freshly-cloned repo bootable into a fully-restored Demos node with both forks pre-applied and a secure consensus shard. End-to-end pristine-machine launch is now `./run --docker --clean`.

- **State-Restore-from-Snapshot** (epic 15): commit a verified snapshot of historical state (13 880 `gcr_main` + 1 382 `gcr_storageprogram`) and teach the genesis builder to restore it under a single transaction. Pre-applies `osDenomination` + `gasFeeSeparation` migrations at block 0 (deterministic, consensus-free). Plan + runbook in `forking/restore/`.
- **Multi-node interop tests** (epic 16): two-node side-by-side via the new `testing/multinode/` overlay. Validated peer discovery, genesis-hash match, sync convergence, cold-companion fast-sync, plus five rejection edge cases (tampered genesis, tampered snapshot, pre-populated PG, missing snapshot, corrupted manifest).
- **Consensus shard enforcement** (epic 17): `getShard()` now intersects online peers with the active baked validator set. Empty-validator fallback gated by `DEMOS_REQUIRE_VALIDATORS=true` for production. Closes the gap revealed by epic 16 P10 where any peer could mine.

## What's in the commit

**Snapshot pipeline**
- `data/snapshot/{gcr_main,gcr_storageprogram,identity_commitments}.jsonl` + `manifest.json` (≈20 MB committed; deterministic regeneration via `bun snapshot:transform`).
- `scripts/state-snapshot/{transform,verify-snapshot,dry-run-restore}.ts` + `package.json` scripts.
- `src/libs/blockchain/genesis/{loadSnapshot,restoreSnapshot,verifySnapshot,seedValidators,applyForksAtGenesis}.ts`.
- `src/libs/blockchain/chainGenesis.ts` branches on `loadSnapshot()` availability; legacy fallback preserved byte-identical.

**Genesis config**
- `data/genesis.json`: `balances: []`, fork `activationHeight: null` (block-N hook stays dormant; `fork_state` is the durable record), real `treasuryAddress` baked, `validators[]` with node2, node3, operator, and a test pubkey.

**Operator workflow**
- `scripts/docker-run --clean` wipes named volumes (`demos_pgdata`, `demos_node_data`) before `up`. Equivalent to bare-metal `./run -b true`.
- `scripts/gen-test-identity.ts` + `scripts/derive-pubkey.ts` for multi-node setups.

**Multi-node harness**
- `testing/multinode/docker-compose.companion.yml` + `up-companion.sh` overlay.

**Consensus shard fix**
- `src/libs/consensus/v2/routines/getShard.ts` intersects online peers with `GCR.getGCRValidatorsAtBlock(lastBlock)`. Empty set → fallback (default loud warning + legacy behavior) or fatal throw if `DEMOS_REQUIRE_VALIDATORS=true`.

**Docs**
- `forking/restore/PLAN.md` + `forking/restore/RUNBOOK.md` (1843 words, 10 sections including §2.4 on validator-set enforcement).
- Cross-link in `forking/RUNBOOK_FORK_ACTIVATION.md` flagging the production env var.

## Quality gates

- **103 pass / 9 skip / 0 fail** across `testing/state-snapshot/`, `testing/forks/restore/`, `testing/consensus/`.
- 9 skips are PG-gated tests that pass on a live Postgres; gate uses `it.skipIf(!pgAvailable)` so CI reports them as proper skips, not silent passes.
- `bun snapshot:verify` exits 0 against committed snapshot.
- Snapshot JSONL sha256s are deterministic on re-run.
- No new TypeScript errors introduced. (Pre-existing `l2ps-messaging` + `verify-release-gate` errors unchanged.)
- `osDenomination` and `gasFeeSeparation` migration code untouched.
- Legacy non-snapshot genesis path preserved byte-identical.

## Test plan

Reviewer checklist:

- [ ] `bun test testing/state-snapshot/ testing/forks/restore/ testing/consensus/` reports `103 pass / 9 skip / 0 fail`.
- [ ] `bun snapshot:verify` exits 0.
- [ ] Re-running `bun snapshot:transform` against the same source produces byte-identical output (sha256 of every file in `data/snapshot/` stable).
- [ ] Fresh-machine smoke: `git clone … && cd node && git checkout chore/restore && ./run --docker --clean -d` then verify:
  - `gcr_main` row count = 13 880
  - `gcr_storageprogram` row count = 1 382
  - `validators` row count = 4
  - `fork_state` has both forks `applied=true, applied_at_block=0`
  - `SELECT SUM(balance) FROM gcr_main` = `15024999999998868586000000000` (post-fork OS)
  - `blocks` advances beyond 0 once consensus tick fires
- [ ] Set `DEMOS_REQUIRE_VALIDATORS=true` in a test container with the `validators` table emptied; node refuses to operate with the documented error message.

## Risks

- Block-0 hash diverges from the previous testnet's genesis. This is intentional (chain wipe) and documented in the RUNBOOK.
- Production environments must set `DEMOS_REQUIRE_VALIDATORS=true`; without it `getShard` falls back to the legacy permissive behavior with a loud warning. The RUNBOOK calls this out.
- Snapshot data is ≈20 MB committed to repo. Acceptable trade for reproducibility; alternative is post-clone fetch from a CDN/IPFS (deferred).

## Out of scope

- Three-validator Byzantine consensus rehearsal — deferred until mnemonics for the other two baked validator pubkeys (`0x24c664…` node3, `0xc8bc58…` node2) are available. Tracked as Mycelium task `#156` (closed/deferred).
- IPFS/CDN distribution of the snapshot.
- TOCTOU between `loadSnapshot` verify and `restoreSnapshot` stream — adversary needs filesystem write between two sequential async ops with operator already trusted; documented but not closed (would need verify-then-buffer pipeline refactor).

## Summary by CodeRabbit

* **New Features**
  * Added snapshot-based state restoration for genesis block initialization with validator seeding and deterministic fork pre-application
  * New CLI tools for snapshot operations: verification, transformation, and dry-run restoration testing

* **Documentation**
  * Added comprehensive operational runbooks and technical plans for snapshot restoration workflows

* **Tests**
  * Extensive test coverage for snapshot restoration, fork activation, and validator seeding

* **Chores**
  * Updated Docker Compose configuration and scripts for multi-node testing scenarios

[Review Change Stack](https://app.coderabbit.ai/change-stack/kynesyslabs/node/pull/825?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: data/genesis.json [hunk 1]
```diff
@@ -7,36 +7,50 @@
...
+          "first_seen": 0,
+          "valid_at": 0
+      },
+      {
+          "address": "0xe6313748a6bbd8e0ebdb578ca0df5f050711819c8dba291930969f72a7f5cdac",
+          "status": "2",
+          "connection_url": null,
+          "staked_amount": "1000000000000000000",
+          "first_seen": 0,
+          "valid_at": 0
```
[the remark this answers is not in the document]
Comment (tcsenpai) [reply data/genesis.json:53]: Verified false positive. The entity `Validators.connection_url` is typed `string | null` (TypeORM `Column("text", { nullable: true })`), and `validateSeed` in `seedValidators.ts` was updated in this iteration to explicitly accept null (or a non-empty string). The two `null` values in `data/genesis.json` are intentional: node5 has no stable public URL, and the 4th baked test validator is used in local multi-node tests only. Resolving.

File: src/libs/blockchain/chainGenesis.ts [hunk 2]
```diff
@@ -102,43 +108,103 @@ export async function generateGenesisBlock(genesisData: any): Promise<Block> {
...
+                await seedValidators(
+                    em,
+                    genesisData.validators as GenesisValidatorSeed[],
+                )
+            }
+            // Pre-apply forks deterministically at genesis. Migration output is
+            // a pure function of input state — no consensus block-1 hook needed.
+            // Pristine boot is fully post-fork; solo nodes don't sit at genesis.
+            await applyForksAtGenesis(em, genesisData.forks)
+        })
```
[the remark this answers is not in the document]
Comment (tcsenpai) [reply src/libs/blockchain/chainGenesis.ts:169]: Addressed in greploop iter 1 via approach (b). Inspected `insertBlock` in chainBlocks.ts — it owns an internal `dataSource.transaction` with savepoint-per-tx, Merkle update, governance hooks. Threading the outer `em` through would require touching `Mempool.removeTransactionsByHashes`, `updateMerkleTreeAfterBlock`, `tallyUpgradeVotes`, `applyNetworkUpgrade`, and `loadNetworkParameters`. Instead, `restoreSnapshot.preflightEmpty` now distinguishes "partial-genesis" (`gcr_main` populated, `blocks` empty) from "populated DB" with case-specific recovery instructions. The atomicity gap is documented in the comment block in chainGenesis.ts. Resolving.
